website/www/site/content/en/blog/unit-testing-blog.md

```diff
@@ -0,0 +1,149 @@
+
+Suppose we write a particular PTransform that reads data from a CSV file, gets passed through a custom function for parsing, and is written back to another Google Cloud Storage bucket (we need to do some custom data formatting to have data prepared for a downstream application).
+
+
+The pipeline is structured as follows:
+
+### Example pipeline 1
+
+    #The following packages are used to run the example pipelines
+    import apache_beam as beam
+    import apache_beam.io.textio.ReadFromText
+    import apache_beam.io.textio.WriteToText
+
+    with beam.Pipeline(argv=self.args) as p:
+    result = p | ReadFromText("gs://my-storage-bucket/csv_location.csv")
+               | beam.ParDo(lambda x: custom_function(x))
+               | WriteToText("gs://my-output-bucket-location/")
+
+We then add a custom function to our code:
+
+    def custom_function(x):
+         ...
+         returned_record = requests.get("http://my-api-call.com")
+         ...
+         if len(returned_record)!=10:
+            raise ValueError("Length of record does not match expected length")
+
+In this scenario, we recommend the following best practices:
+
+1. You don’t need to write any unit tests for the already supported connectors in the Beam Library, such as ReadFromBigQuery and WriteToGCS. These connectors are already tested in Beam’s test suite to ensure correct functionality.
+2. You should write unit tests for any custom operations introduced in the pipeline, such as Map, FlatMap, Filter, ParDo, and so on. We recommend adding tests for any lambda functions utilized within these Beam primitives. Additionally, even if you’re using a predefined function, treat the entire transform as a unit, and test it.
+
+Let’s use the following pipeline as an example. Because we have a function, we should write a unit test to ensure that our function works as intended.
+
+    def compute_square(element):
+        return element**2
+
+    with beam.Pipeline(argv=self.args) as p1:
+        result = p1 | ReadFromText("gs://my-storage-bucket/csv_location.csv")
+                    | beam.Map(compute_square)
+                    | WriteToText("gs://my-output-bucket-location/")
+
+
+Now let’s use the following pipeline as another example. Because we use a predefined function, we don’t need to unit test the function, as `str.strip`, is tested elsewhere. However, we do need to test the output of the `beam.Map` function.
+
+    with beam.Pipeline(argv=self.args) as p2:
+        result = p2 | ReadFromText("gs://my-storage-bucket/csv_location.csv")
+                    | beam.Map(str.strip)
+                    | WriteToText("gs://my-output-bucket-location/")
+
+
+Here are the corresponding tests for both pipelines:
```

Comment:
Do these codes actually run?

Reply:
You cannot run them in the browser, these are meant to be snippets for example purposes.